backend/src/main/kotlin/app/ehrenamtskarte/backend/matomo/Matomo.kt

```diff
@@ -0,0 +1,159 @@
+        if (projectConfig.matomo == null) return
+        val staticCards = cards.filter { it.codeType === CodeType.STATIC }
+        val dynamicCards = cards.filter { it.codeType === CodeType.DYNAMIC }
+        if (staticCards.isNotEmpty() && dynamicCards.isNotEmpty()) {
+            this.sendBulkTrackingRequest(
+                projectConfig.matomo,
+                listOf(
+                    buildCardsTrackingRequest(request, staticCards.first().regionId, query, CodeType.STATIC, staticCards.count()),
+                    buildCardsTrackingRequest(request, staticCards.first().regionId, query, CodeType.DYNAMIC, dynamicCards.count())
+                )
+            )
+        } else if (staticCards.isNotEmpty()) {
+            this.sendTrackingRequest(projectConfig.matomo, buildCardsTrackingRequest(request, staticCards.first().regionId, query, CodeType.STATIC, staticCards.count()))
+        } else if (dynamicCards.isNotEmpty()) {
+            this.sendTrackingRequest(
+                projectConfig.matomo,
+                buildCardsTrackingRequest(request, dynamicCards.first().regionId, query, CodeType.DYNAMIC, dynamicCards.count())
+            )
+        }
+    }
+
+    fun trackVerification(projectConfig: ProjectConfig, request: HttpServletRequest, query: String, cardHash: ByteArray, codeType: CodeType, successful: Boolean) {
+        if (projectConfig.matomo === null) return
+        val card = transaction { CardRepository.findByHash(projectConfig.id, cardHash) }
+        this.sendTrackingRequest(
+            projectConfig.matomo,
+            MatomoRequest.builder()
+                .eventAction(query)
+                .eventCategory(EventCategory.Cards.toString())
+                .eventName(codeType.toString())
+                .eventValue(if (successful) 1 else 0)
+                .customTrackingParameters(if (card != null) mapOf("dimension1" to card.regionId) else emptyMap())
+                .also { attachRequestInformation(it, request) }
+        )
+    }
+
+    fun trackActivation(projectConfig: ProjectConfig, request: HttpServletRequest, query: String, cardHash: ByteArray, successful: Boolean) {
+        if (projectConfig.matomo === null) return
+        val card = transaction { CardRepository.findByHash(projectConfig.id, cardHash) }
+        this.sendTrackingRequest(
+            projectConfig.matomo,
+            MatomoRequest.builder()
+                .eventAction(query)
+                .eventCategory(EventCategory.Cards.toString())
+                .eventValue(if (successful) 1 else 0)
+                .customTrackingParameters(if (card != null) mapOf("dimension1" to card.regionId) else emptyMap())
+                .also { attachRequestInformation(it, request) }
+        )
+    }
+
+    fun trackSearch(projectConfig: ProjectConfig, request: HttpServletRequest, query: String, params: SearchParams, numResults: Int) {
+        if (projectConfig.matomo === null) return
+        if (params.categoryIds != null) return
```

Comment:
Why do you reteurn when categoryIds is not null?
Maybe, we should use some special encoding for the matomo search query?

Reply:
See https://github.com/matomo-org/matomo-java-tracker/issues/133. I decided to add the categories in the search query. Do you think it makes more sense to add the plain text category names here? Would be a little bit more effort, but easier to evaluate.